### DISCUSSION TITLE
Quarkus using eventloop thread for Kotllin suspend functions

### DISCUSSION THREAD

**Discussion by mpienaar-t** [quarkusio/quarkus#32385] [Q&A] [answered]
Hello Friends!

I've been using Quarkus with Kotlin in an application that is destined for production, with the quarkus-kotlin extension installed, and I noticed that I was getting a warning sometimes about the eventloop thread being blocked for more than 2 seconds.
This was of course a concern to me. My understanding, after reading https://quarkus.io/blog/resteasy-reactive-smart-dispatch/ was that resource endpoints should automatically offload the work to a worker or IO thread, depending on circumstances mentioned in the document. I created a new app using https://quarkus.io/guides/kotlin and experimented a bit. I noticed that if your endpoint function is a suspend function, then if you print out the thread's name it is an eventloop thread.
eg.
```
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    suspend fun hello(): String {
        println("Thread Name: ${Thread.currentThread().name}") // Prints out "Thread Name: vert.x-eventloop-thread-0"
        return "Hello from RESTEasy Reactive"
    }
```

If you remove the suspend keyword, then the thread's name is an executor thread.
eg. 
```
    @GET
    @Produces(MediaType.TEXT_PLAIN)
    fun hello(): String {
        println("Thread Name: ${Thread.currentThread().name}") // Prints out "Thread Name: executor-thread-0"
        return "Hello from RESTEasy Reactive"
    }
```

Is this behaviour expected? I am guessing that this is the cause of the occasional "eventloop thread being blocked" warning.

Thanks so much for your help.

[a comment by an automated account is not shown]

**Comment by geoand** [comment, marked as answer]
Hi, 

Yes, what you are seeing with the thread names is the expected behavior

**Reply by mpienaar-t** [reply]
Thanks for the info, @geoand . Much appreciated. What is the suggested way to avoid the warning about running long processes on the event loop thread in this case?

**Reply by geoand** [reply]
You need to avoid performing any long running, operations in those methods

**Reply by mpienaar-t** [reply]
If this endpoint makes a database/api call or something that could take long - where would you do it? Is there a reference example for how to handle these situations?

**Reply by geoand** [reply]
In that case you probably either want to not have the function be a `suspend` function or use Kotlin's `runBlocking`

**Reply by mpienaar-t** [reply]
Ok thanks so much. I really appreciate your quick responses.

**Reply by geoand** [reply]
You are very welcome.
